Favorites button addition + other fixes

blackletum opened the pull request:
These changes add a favorites button that allows you to toggle your favorites list off/on.

![Screenshot of favorites checkbox](https://i.imgur.com/x6KSZeX.png)

![Screenshot of favorites toggled](https://i.imgur.com/YPGbQb1.png)

To-do:

- Button (or the entire program) should somehow show that Favorites are being shown in an easy-to-understand way. The star symbol may be the ticket for this. 
- There is code to add a star symbol to those in the favorites list that I didn't get working bc I shoved it in at 5:30 am, but it should be shown to the left of the game name in the list maybe?
- Right clicking on an object in the program should select that program

## Summary by Sourcery

Add favorites functionality to the game launcher, allowing users to mark and filter games as favorites

New Features:
- Implement a favorites toggle button to show/hide favorite games
- Add a favorite checkbox in the game editor
- Display a star symbol for favorite games in the game list

Enhancements:
- Modify game list rendering to support favorite game highlighting
- Update game data model to include favorite status
- Improve game list filtering based on favorite status

blackletum commented:
Just noticed a _huge_ issue, when you toggle favorites it still tries to select whatever was first/second/etc in the list instead of looking at the new list...

so in my case when I added Pre-Fortress 2 to my favorites, it was the 2nd game, and when I select it after clicking Toggle Favorites, it selects Cyber Diver.

PracticeMedicine reviewed (approved):
LGTM!

y2k04 commented:
Nicely done! I will look at the code more and do my own testing. After that, I will definitely be pushing this.

Thank you for contributing! I very much appreciate it :3

y2k04 commented:
@sourcery-ai review

y2k04 commented:
Also trying out Sourcery to see how it is

blackletum commented:
So far from using my version, I've never seen over 17 mb of memory usage in task manager. Obviously this isn't comprehensive whatsoever, but I slapped together a quick ps1 script together to track usage. It tracked memory usage of the application once per second.

I then opened up the application and started toggling favorites on/off, adding things to favorites back and forth, added a new game, deleted a game, starting a new game then closing it to record time, etc. Extracting the information from that ps1 log export and averaging it out to 33.05 MB.

So I think that might be about the worst you'd see as far as RAM usage, at least best I can tell.

y2k04 commented:
> So far from using my version, I've never seen over 17 mb of memory usage in task manager. Obviously this isn't comprehensive whatsoever, so I slapped together a quick ps1 script together to track usage. It tracked memory usage of the application once per second.
> 
> I then opened up the application and started toggling favorites on/off, adding things to favorites back and forth, added a new game, deleted a game, starting a new game then closing it to record time, etc. Extracting the information from that ps1 log export and averaging it out resulted in a result of 33.05 MB. I also just checked the median and it was 33.6 MB.
> 
> So I think that might be about the worst you'd see as far as RAM usage, at least best I can tell.
> 
> Obviously, however, if you can fix up the code and make it use less RAM, that's ideal and should be pursued.

Defs. Now that my computer decided to stop dying, I'm gonna download your fork finally :D

y2k04 commented:
What do you think of the changes I added?

PracticeMedicine reviewed (approved):
The changes looks good to me, but optionally you should also make use of `using` statements or declarations for the `stream`, `reader`, `writer` and other variables implementing `IDisposable` instead of storing them so that it's disposed till its used again in other functions. Again, it's optional since it's lightweight enough but it's your decision whether you want to implement these changes.

y2k04 commented:
My reasoning for not using the `using` function in that situation is because a few functions write to the `config.json` and it doesn't seem useful to continuously reopen a handle when it could just stay open for the lifecycle of the app.

blackletum commented:
Compiled, looks good to my eyes.
Only weird issues I ran into is that one of my games seemed to be doubled, but I'm not sure when that appeared or how it occurred. Figured I'd share it here though regardless. I wonder if there was an issue when making it a favorite that it saved 2 entries? (I cannot repeat this issue)

Another thing I'd put on the to-do that I just thought of is having the window remember the resize when it opens next, I have some games with longer names that get cut off and also when I resize the text window, the "toggle favorites" button gets shoved off the side.

Not directly related to the project: did **sourcery** help you out any?

y2k04 commented:
> Compiled, looks good to my eyes. Only weird issues I ran into is that one of my games seemed to be doubled, but I'm not sure when that appeared or how it occurred. Figured I'd share it here though regardless. I wonder if there was an issue when making it a favorite that it saved 2 entries? (I cannot repeat this issue)
> 
> Another thing I'd put on the to-do that I just thought of is having the window remember the resize when it opens next, I have some games with longer names that get cut off and also when I resize the text window, the "toggle favorites" button gets shoved off the side.
> 
> Not directly related to the project: did **sourcery** help you out any?
> 
> Edit: I'm having an issue now actually where if I add new favorites, some of my pre-existing favorites not only get unfavorited (probably) but they also go missing from my game list entirely. It seems to happen sporadically, though...
> 
> edit 2: Another nice-to-have would be if the program could remember if favorites was toggled or not before you close, now that I think about it...

I wonder if this part of the addGameButton_Click function had anything to do with it 🤔
![image](https://github.com/user-attachments/assets/61d14c51-fafb-40a2-bf84-7e092461d6c5)

I think adding the window size to the config.json would be good. The toggle favourites button can be fixed the same way I made the Play button stay in place (docking and padding).

Tbh, sourcery did not. I will remove it from my account after finishing writing this.

That is odd that they are disappearing.

I will work on GameLauncher more in a week because of my full-time job. Feel free to add, change or fix things while I am gone :)

PracticeMedicine commented:
> My reasoning for not using the `using` function in that situation is because a few functions write to the `config.json` and it doesn't seem useful to continuously reopen a handle when it could just stay open for the lifecycle of the app.

Understandable, however this approach has *drawbacks* like file locking, concurrency issues and resource leaks that you should consider thinking about it.

<sub>P.S: if Game Launcher crashes, sometimes the handle (in this case `StreamReader`) might not be properly released so the file will be remain locked until the OS cleans it up unless a different `FileShare` parameter in the `File.Open` function is used. If you still don't want to repeatedly opening and closing the file, consider making a singleton pattern or implement a proper disposal.</sub>

y2k04 commented:
By the way, last night I realised what the AI meant by storing the Game object in the Tag field. It was to replace the use of an index! That would make it easier to handle and make the `visibleGames` object deprecated and could be removed.

y2k04 commented:
Forgot to add to the changes that the Config now contains a favourites toggle and should be enabled (if toggled) when relaunching the app. I forgot to test that though, as well as if the favourites toggle changes colour if it is enabled on start up.

PracticeMedicine reviewed (commented):
i'll test them out later on, but overall looks good to me also.

PracticeMedicine commented on Launcher.cs:
```diff
@@ -43,13 +43,29 @@ private async void SetupLauncher()
 
-            var data = await reader.ReadToEndAsync();
-            if (data != "[]" && data != string.Empty)
+            var rawData = await reader.ReadToEndAsync();
```
This requires `null` checking, here is a simple change I did:
```cs
...
var rawData = await reader.ReadToEndAsync();
// PracticeMedicine --
// if the raw data is null then initialize a new Config and quickly abort this function
if (string.IsNullOrEmpty(rawData))
{
    config = new()
    {
        Games = new(),
        FavouritesToggled = false
    };
    return;
}
...
```

y2k04 commented:
I am going to pull this and create a new release now.

Thank you @blackletum and @PracticeMedicine for your help! ^w^
